kong/observability/logs.lua

```diff
@@ -0,0 +1,179 @@
+local _M = {
+  maybe_push = function() end,
+  get_request_logs = function() return {} end,
+  get_worker_logs = function() return {} end,
+}
+
+if ngx.config.subsystem ~= "http" then
+  return _M
+end
+
+
+local request_id_get = require "kong.observability.tracing.request_id".get
+local time_ns = require "kong.tools.time".time_ns
+local deep_copy = require "kong.tools.utils".deep_copy
+
+local get_log_level = require "resty.kong.log".get_log_level
+local constants_log_levels = require "kong.constants".LOG_LEVELS
+
+local table_new = require "table.new"
+local table_clear = require "table.clear"
+local string_buffer = require "string.buffer"
+
+local ngx = ngx
+local kong = kong
+local table = table
+local tostring = tostring
+
+local ngx_null = ngx.null
+local table_pack = table.pack -- luacheck: ignore
+
+local MAX_WORKER_LOGS = 1000
+local MAX_REQUEST_LOGS = 1000
+local NGX_CTX_REQUEST_LOGS_KEY = "o11y_logs_request_scoped"
+
+local worker_logs = table_new(100, 0)
+local logline_buf = string_buffer.new()
+
+
+-- WARNING: avoid using `ngx.log` in this function to prevent recursive loops
+local function configured_log_level()
+  local ok, level = pcall(get_log_level)
+  if not ok then
+    -- This is unexpected outside of the context of unit tests
+    local level_str = kong.configuration.log_level
+    _G.native_ngx_log(ngx.WARN,
+      "[observability] OpenTelemetry logs failed reading dynamic log level. " ..
+      "Using log level: " .. level_str .. " from configuration."
+    )
+    level = constants_log_levels[level_str]
+  end
+
+  return level
+end
+
+
+-- needed because table.concat doesn't like booleans
+local function concat_tostring(tab)
+  local tab_len = #tab
+  if tab_len == 0 then
+    return ""
+  end
+
+  for i = 1, tab_len do
+    local value = tab[i]
+
+    if value == ngx_null then
+      value = "nil"
+    else
+      value = tostring(value)
```

Comment:
Ah, Apologies, I forgot to mention that `ngx.log` only accepts `number`, `boolean`, `string`, `table with __tostring`, and `ngx.null`.

To keep the consistency, shall we skip generating OTel logs for this case as `ngx.log` will panic?

https://github.com/openresty/lua-nginx-module/blob/6477a7b46d78e9ad841f2f3b2e1d37161b748669/src/ngx_http_lua_log.c#L189
https://github.com/openresty/lua-nginx-module/blob/6477a7b46d78e9ad841f2f3b2e1d37161b748669/src/ngx_http_lua_log.c#L171

Reply:
see my other comment ^